### PR TITLE
Merge QGB feature branch into main

### PR DESCRIPTION
**Pull request by SweeXordious**
This merge will add the QGB state machine to Celestia-app.

### CODE REVIEW

**Comment by SweeXordious** [comment]
I will update the QGB universal nonce ADR to make the review easier

**Comment by evan-forbes** [comment]
> I will update the QGB universal nonce ADR to make the review easier

yeah we need to straighten out all of the ADR numbers

File: scripts/protocgen.sh
Comment (evan-forbes) [inline scripts/protocgen.sh:31]: agree with @rootulp that we might need to reopen or create a new issue to make sure we don't forget about this

File: x/qgb/keeper/keeper_data_commitment.go
Comment (rootulp) [inline x/qgb/keeper/keeper_data_commitment.go:10]:
```suggestion
// GetCurrentDataCommitment creates latest data commitment at current height according to
```

File: scripts/single-node.sh
Comment (evan-forbes) [inline scripts/single-node.sh:19]: @YazzyYaz just an fyi that this change requires validators to add two extra fields when they create a validator. This differs from the normal cosmos-sdk, so if we have docs for creating a validator, we will need to update them.

also @SweeXordious we also want to make a note of this somewhere to make sure that we document this in our own docs.

File: x/qgb/keeper/keeper_valset.go
Comment (rootulp) [inline x/qgb/keeper/keeper_valset.go:46]:
```suggestion
// SetLastUnBondingBlockHeight sets the last unbonding block height. Note this
// value is not saved to state or loaded at genesis. This value is reset to zero
// on chain upgrade.
```

File: x/qgb/keeper/keeper_valset.go
Comment (rootulp) [inline x/qgb/keeper/keeper_valset.go:53]:
```suggestion
// GetLastUnBondingBlockHeight returns the last unbonding block height or zero
// if not set. This value is not saved or loaded at genesis. This value is reset
// to zero on chain upgrade.
```

File: testutil/common.go
Comment (evan-forbes) [inline testutil/common.go:75]: [todo] we should remove this and just use the apps or if that's not possible then we should use the testapp's

File: testutil/common.go
Comment (evan-forbes) [inline testutil/common.go:470]: [blocking][new issue needed]

we should ideally consolidate this with https://github.com/celestiaorg/celestia-app/blob/e18de882c3de4ee92e4632af21d4007c1b642adb/testutil/test_app.go#L71

File: x/qgb/keeper/query_valset.go
Comment (rootulp) [inline x/qgb/keeper/query_valset.go:12]:
```suggestion
// LastValsetRequestBeforeNonce queries the last valset request before nonce
```

File: x/qgb/handler.go
Comment (evan-forbes) [inline x/qgb/handler.go:13]: [question]

just double checking, do we have an issue for this?

File: x/qgb/types/validator.go
Comment (evan-forbes) [inline x/qgb/types/validator.go:36]: [comment]
we might want to change the name of these to avoid confusion, but its a very low priority since they're strictly internal and we inherited them from the gravity module

**Review by evan-forbes — COMMENTED** [review]
super super super dope. glad we're finally getting this merged, and I'm really proud of us (almost 100% @SweeXordious lol) for iterating and shipping so fast. Not only are we shipping a working version of the qgb, but one that has altered the original design dramatically to produce something significantly better. 

Really big hats off, @SweeXordious. Top shelf work.

File: docs/architecture/ADR-003-QGB-DataCommitment.md
Comment (SweeXordious) [reply docs/architecture/ADR-003-QGB-DataCommitment.md:23]: I am planning to edit the universal nonce ADR to say that the design defined in previous  ADRs was changed and we no longer store the confirms in state.  
What do you think about that?

File: x/qgb/types/genesis.go
Comment (rootulp) [inline x/qgb/types/genesis.go:44]: remove duplicate word

```suggestion
// ParamSetPairs implements the ParamSet interface and returns all the key/value
// pairs of auth module's parameters.
```

File: scripts/single-node.sh
Comment (SweeXordious) [reply scripts/single-node.sh:19]: We should decide on which docs do we want to write for now. Do we want to wait until the orchestrator/relayer are ready and document everything? or we want to update existing validator docs?

File: testutil/common.go
Comment (SweeXordious) [reply testutil/common.go:470]: Should be done before or after the merge?
I guess it can be done after the merge.

File: x/qgb/handler.go
Comment (SweeXordious) [reply x/qgb/handler.go:13]: we have this: https://github.com/celestiaorg/celestia-app/issues/684 
Should be detailed more, but I guess it summarizes everything until we create a more concrete tracking issue with explicit tasks

File: scripts/single-node.sh
Comment (YazzyYaz) [reply scripts/single-node.sh:19]: @evan-forbes does this apply to Mamaki users or the new Arabica testnet?

File: x/qgb/types/validator.go
Comment (SweeXordious) [reply x/qgb/types/validator.go:36]: Do you have any other name in mind?  I guess there are lot of structs that  can be renamed to self  reflect what  they do. Do we  need an issue?

File: go.mod
Comment (SweeXordious) [reply go.mod:179]: It was used in the orchestrator. We  can remove the dependency I guess and reintroduce it later. What do you think?

File: x/qgb/types/valset.go
Comment (rootulp) [inline x/qgb/types/valset.go:58]:
```suggestion
// Hash mimics the 'computeValsetHash' function used by the qgb contracts by using
```

File: x/qgb/types/valset_confirm.go
Comment (rootulp) [inline x/qgb/types/valset_confirm.go:35]: [question] are there naming conventions around the return value? `/qgb.MsgValsetConfirm` isn't consistent with the previous type `data_commitment_confirm` but it's not clear to me which naming conventions are preferred for this value.

File: docs/architecture/ADR-003-QGB-DataCommitment.md
Comment (evan-forbes) [reply docs/architecture/ADR-003-QGB-DataCommitment.md:23]: cool, makes sense

**Review by rootulp — COMMENTED** [review]
I did a first pass review prior to Thursday's synchronous review. A lot of my feedback pertains to code comments / consistency / naming. They can be addressed in follow-up PRs if we want to minimize the changes in this PR.

File: go.mod
Comment (evan-forbes) [reply go.mod:179]: yeah, let's remove it until we need it

File: scripts/single-node.sh
Comment (evan-forbes) [reply scripts/single-node.sh:19]: > does this apply to Mamaki users or the new Arabica testnet?

this will apply to whichever chains will be running using v0.7.0 of celestia-app.

File: testutil/common.go
Comment (evan-forbes) [reply testutil/common.go:470]: yeah, after, but we should create an issue for it so we don't forget

File: scripts/single-node.sh
Comment (YazzyYaz) [reply scripts/single-node.sh:19]: OK cc @sysrex @Bidon15 just a heads up for when version 0.7.0 of celestia-app is released, we will need to update the configs on running Arabica.

File: x/qgb/types/validator.go
Comment (evan-forbes) [reply x/qgb/types/validator.go:36]: no, I definitely don't think we have to change this in this PR or even create an issue. It was just a comment to see if something else pops up.

File: scripts/single-node.sh
Comment (YazzyYaz) [reply scripts/single-node.sh:19]: https://github.com/celestiaorg/docs/issues/222

File: scripts/protocgen.sh
Comment (SweeXordious) [reply scripts/protocgen.sh:31]: https://github.com/celestiaorg/celestia-app/issues/744

File: go.mod
Comment (evan-forbes) [reply go.mod:179]: yeah, https://github.com/celestiaorg/celestia-app/pull/742#discussion_r975796064

we can remove. We might have to come up with a solution in the future when we merge the relayer/orchestrator.

perhaps we can create a single cleanup issue @SweeXordious , and then follow up after we merge this PR given that we want to not squash this.

**Comment by evan-forbes** [comment]
if we don't want to create a cleanup issue @SweeXordious with all the feedback here (sounds kinda tedious), we could create a different PR with all the feed back and squash it into this PR

**Comment by SweeXordious** [comment]
@evan-forbes Yes, you're right. I will apply the changes on a separate PR and merge it here.

File: testutil/common.go
Comment (SweeXordious) [reply testutil/common.go:470]: https://github.com/celestiaorg/celestia-app/issues/745

File: proto/qgb/query.proto
Comment (SweeXordious) [reply proto/qgb/query.proto:22]: Queries for the attestations requests.  An attestation request is a stored attestation waiting to be signed by an orchestrator. 
What do you think about: `queries for attestations requests waiting to be signed by an orchestrator`?

File: testutil/common.go
Comment (SweeXordious) [reply testutil/common.go:423]: The `Unhandled error `  is not flagged?

File: testutil/common.go
Comment (SweeXordious) [reply testutil/common.go:1]: Will be fixed in here: https://github.com/celestiaorg/celestia-app/issues/745

File: x/qgb/abci.go
Comment (SweeXordious) [reply x/qgb/abci.go:67]: makes sense :+1:

File: scripts/protocgen.sh
Comment (SweeXordious) [reply scripts/protocgen.sh:29]: https://github.com/celestiaorg/celestia-app/issues/744

File: x/qgb/types/valset_confirm.go
Comment (SweeXordious) [reply x/qgb/types/valset_confirm.go:35]: There is an issue to decide on conventions for this and also on the cosmos-sdk side:
https://github.com/celestiaorg/celestia-app/issues/303 and https://github.com/celestiaorg/cosmos-sdk/issues/136

File: scripts/single-node.sh
Comment (Bidon15) [reply scripts/single-node.sh:19]: Bumping our discussion, 
What exact rc candidate of `0.7.0` do we want to deploy in arabica for celestia-app?

Comment (evan-forbes) [reply scripts/single-node.sh:19]: depends on https://github.com/celestiaorg/celestia-node/pull/1147, but if we can't merge that, then I would recommend sticking to v0.7.0-rc-1

**Comment by evan-forbes** [comment]
if we can't figure out this silly CI bug tmrw morning, I'm calling it as this thing is stupid and weirdly frustrating. We have more important things to do atm lol

**Comment by SweeXordious** [comment]
@evan-forbes this commit seems to fix it: https://github.com/celestiaorg/celestia-app/pull/742/commits/13a9db9f9cf43a593fa98f7cfc09fdcba20a2277
It updates `tendermintdev/sdk-proto-gen` to `v0.7`. For some reason, the `latest` is still pointing towards `v0.6` which doesn't have `go1.18` installed.

Also, it removes this https://github.com/celestiaorg/celestia-app/blob/d53f66b67946d20af5a24487c2f79da4ebb1b171/scripts/protocgen.sh#L11 from `protocgen.sh`. Is there a reason why we have? it seems that it's already imported in the go.mod file.

**Comment by evan-forbes** [comment]
> Is there a reason why we have? it seems that it's already imported in the go.mod file.

it does seem redundant, so I'm good with removing it.

great work debugging!

**Review by evan-forbes — APPROVED** [review]

**Comment by SweeXordious** [comment]
Finally mergiiiiing :tada: :tada: 
Thanks a lot for the reviews and for all the help provided :+1: :+1:

**Pull request merged**
